Feature/convert planner service to action

woodaaron opened the pull request:
Addresses #154.
Converts "process planner" calls from a service to an action. This allows the Rviz to continue operation even while planning is happening. Feedback is provided as a string which indicates the last step completed (e.g. generated scan path for surface 0).

Commit:
Created action server for process path planning

Commit:
Removed ProcessPlanning.srv and PathPlanning.action (unused)
Simplified action types for ProcessPlanning.action

Commit:
Added ProcessPlanning action client to admin gui

Commit:
Fixed orphaned ProcessPlanning messages

Commit:
Updated simple gui to use the process planning action (rather than service)
Removed blending options from gui and added path planning options

Jmeyer1292 commented:
@woodaaron I pulled in #163 and introduced some conflicts. Please resolve.

AustinDeric commented:
@woodaaron, i think we just need to use the new .travis.yml for this commit.  rosdep is failing, which it should:

```
ERROR: the following packages/stacks could not have their rosdep keys resolved
to system dependencies:
keyence_experimental: Cannot locate rosdep definition for [libsocket]
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Function rosdep_install returned with code '1' after 0 min 1 sec
```

Jmeyer1292 commented on godel_plugins/include/godel_plugins/widgets/robot_blending_widget.h:
```diff
@@ -61,12 +62,18 @@ const std::string GET_AVAILABLE_MOTION_PLANS_SERVICE = "get_available_motion_pla
 public:
-  RobotBlendingWidget(std::string ns = "");
+  RobotBlendingWidget(std::string ns = "") :
+    param_ns_(ns), process_planning_action_client_(PROCESS_PLANNING_ACTION_SERVER_NAME, true)
```
Does this action client needs its own thread (the `true` argument to the second parameter)?

woodaaron replied:
I can, in a reproducible manner, get the simple gui to hang when the client does not have its own thread. I haven't done much testing with the admin GUI, but decided to keep the initialization similar between the two types.

Jmeyer1292 commented on godel_msgs/action/ProcessPlanning.action:
```diff
@@ -0,0 +1,19 @@
+# Goal
+int32 GENERATE_MOTION_PLAN_AND_PREVIEW=1
+int32 PREVIEW_TOOL_PATH=2
```
This can totally be in a separate PR, but we should probably remove `PREVIEW_TOOL_PATH`: I don't think it does anything anymore. It's an artifact from before we even had a robot in Godel.

woodaaron replied:
Make an issue and I'll address it separately.

Jmeyer1292 commented on godel_surface_detection/include/services/surface_blending_service.h:
```diff
@@ -124,8 +136,15 @@ class SurfaceBlendingService
-  bool process_path_server_callback(godel_msgs::ProcessPlanning::Request& req,
-                                    godel_msgs::ProcessPlanning::Response& res);
+
+  void process_planning_action_callback(const godel_msgs::ProcessPlanningGoalConstPtr &goal);
```
I don't mean to be anal, but all the functions you added except this one follow the ROS naming standard.

Jmeyer1292 commented on godel_process_path_generation/src/process_path_generator_node.cpp:
```diff
@@ -26,7 +26,7 @@
 #include <godel_process_path_generation/utils.h>
 #include <godel_process_path_generation/VisualizeBlendingPlan.h>
-#include <godel_msgs/ProcessPlanning.h>
+#include <godel_msgs/ProcessPlanningAction.h>
```
What's going on in this file? Changes just to keep it compiling? Is this code dead?

woodaaron replied:
Process Planning is now an action as opposed to a service. Converted the header.

Jmeyer1292 commented on godel_simple_gui/src/options/blend_tool_configuration.cpp:
```diff
@@ -1,37 +0,0 @@
-
-void godel_simple_gui::BlendingPlanConfigWidget::update_internal_fields()
-{
-  params_.spindle_speed = ui_->LineEditSpindleSpeed->text().toDouble();
```
I know this may overlap with other changes we've already made... but these parameters are going to be coming back fairly soon. I'm not advocating leaving in dead code for the moment... just an FYI.

Jmeyer1292 commented on godel_simple_gui/src/states/planning_state.cpp:
```diff
@@ -17,10 +15,8 @@ void godel_simple_gui::PlanningState::onStart(BlendingWidget& gui)
-  planning_client_ =
-      gui.nodeHandle().serviceClient<godel_msgs::ProcessPlanning>(PROCESS_PATH_SERVICE);
-  //QtConcurrent::run(this, &PlanningState::makeRequest, gui.options().planning_params);
+  //planning_client_ = gui.nodeHandle().serviceClient<godel_msgs::ProcessPlanning>(PROCESS_PATH_SERVICE);
```
Dead code?

Jmeyer1292 commented on godel_surface_detection/include/services/surface_blending_service.h:
```diff
@@ -81,7 +87,13 @@ struct ProcessPlanResult
 {
 public:
-  SurfaceBlendingService();
+  SurfaceBlendingService() : publish_region_point_cloud_(false),
```
Unless this needs to be in the header, I would prefer the constructor to be in the cpp file.

Jmeyer1292 commented on godel_surface_detection/src/services/surface_blending_service.cpp:
```diff
@@ -1003,6 +991,8 @@ void SurfaceBlendingService::visualizePaths()
 
   tool_path_markers_pub_.publish(path_visualization);
+  process_planning_result_.succeeded = true;
+  process_planning_server_.setSucceeded(process_planning_result_);
```
Have you considered lifting the action code out of `visualizePaths()` and into the handler callback? It seems a little non-intuitive to look at a visualization function and see a code related to some state which may or may not be active.

woodaaron replied:
I was looking at what the last step in the request was "generate_motion_plan_and_PREVIEW", but I agree, it's more visible and makes more sense in the handler callback.

Commit:
Moved action client initializers from header to source files
Fixed naming convention issue
Moved process planning result action success from visualize to callback

Jmeyer1292 commented:
Thanks for your contribution
